packages/react-aria-components/src/Tree.tsx

```diff
@@ -0,0 +1,547 @@
+    children: childRows,
+    items: childItems
+  });
+
+  // Combine the renderChildren and children so both Content and nested TreeItems are properly added to fake DOM and thus added to the built collection
+  return useSSRCollectionNode('item', props, ref, null, [rowContent, collectionChildren]);
+}
+
+/**
+ * A TreeItem represents an individual item in a Tree.
+ */
+const _TreeItem = /*#__PURE__*/ (forwardRef as forwardRefType)(TreeItem);
+export {_TreeItem as TreeItem};
+
+export interface TreeItemContentRenderProps extends ItemRenderProps {
+  // Whether the tree row is expanded.
+  isExpanded: boolean,
+  // Whether the tree row has child rows.
+  hasChildRows: boolean,
+  // What level the tree row has within the tree.
+  level: number
+}
+
+// The TreeItemContent is the one that accepts RenderProps because we would get much more complicated logic in TreeItem otherwise since we'd
+// need to do a bunch of check to figure out what is the Content and what are the actual collection elements (aka child rows) of the TreeItem
+export interface TreeItemContentProps extends Pick<RenderProps<TreeItemContentRenderProps>, 'children'> {}
+
+// TODO does this need ref or context? Its only used to shallowly render the Content node... If it was a more generic collection component then I could see an argument for it
+// having those
+function TreeItemContent(props: TreeItemContentProps) {
+  let ref = useRef(null);
+  let shallow = useShallowRender('content', props, ref);
+  if (shallow) {
+    return shallow;
+  }
+}
+
+const _TreeItemContent = /*#__PURE__*/ TreeItemContent;
+export {_TreeItemContent as TreeItemContent};
+
+function TreeRow<T>({item}: {item: Node<T>}) {
+  let state = useContext(TreeStateContext)!;
+  let ref = useObjectRef<HTMLDivElement>(item.props.ref);
+  let {rowProps, gridCellProps, ...states} = useTreeGridListItem({node: item}, state, ref);
+  let stringFormatter = useLocalizedStringFormatter(intlMessages, 'react-aria-components');
+  let isExpanded = rowProps['aria-expanded'] === true;
+  let hasChildRows = [...state.collection.getChildren!(item.key)]?.length > 1;
+  let level = rowProps['aria-level'];
+
+  let {hoverProps, isHovered} = useHover({
+    isDisabled: !states.allowsSelection && !states.hasAction
+  });
+
+  let {isFocusVisible, focusProps} = useFocusRing();
+  let {checkboxProps} = useGridListSelectionCheckbox(
+    {key: item.key},
+    state
+  );
+
+  let props: TreeItemProps<unknown> = item.props;
+  let renderPropValues = React.useMemo(() => ({
+    ...states,
+    isHovered,
+    isFocusVisible,
+    isExpanded,
+    hasChildRows,
+    level,
+    selectionMode: state.selectionManager.selectionMode,
+    selectionBehavior: state.selectionManager.selectionBehavior
+  }), [states, isHovered, isFocusVisible, state.selectionManager, isExpanded, hasChildRows, level]);
+
+  let renderProps = useRenderProps({
+    ...props,
+    id: undefined,
+    children: item.rendered,
```

Comment:
do tree items support render props? seems like that could conflict with the function for providing children?

Reply:
oh hm, I hadn't really considered that, in my mind I figured users would use the renderProps that TreeItemContent provides. I'll see if I can get rid of the renderProps function for TreeItem entirely

Reply:
Ah, actually I've already have the types setup such that the children of the TreeItem doesn't include the renderProps children definition. I've gotten rid of the `((item: T) => ReactElement)` part of it since users should use a `Collection` for the dynamic case